### PR TITLE
fix(runtime): disallow deletion of account with large state

### PR DESCRIPTION
**Pull request by bowenwang1996**
For account whose state is large, deleting them is not cheap and this PR prevents accounts with more than 10kb of state (excluding contracts) from being deleted. Contract developers should develop their own logic to delete part of the state if needed.

Test plan
---------
* `test_delete_account_too_large`
* `test_delete_account_with_contract_and_small_state`
* `test_delete_account_with_contract_and_large_state`
*  nayduck

### CODE REVIEW

File: runtime/runtime/src/actions.rs [hunk 1]
```diff
@@ -445,7 +446,28 @@ pub(crate) fn action_delete_account(
...
+        >= STABLE_PROTOCOL_FEATURES_TO_VERSION_MAPPING[&ProtocolFeature::DeleteActionRestriction]
+    {
+        let account = account.as_ref().unwrap();
+        let mut account_storage_usage = account.storage_usage;
+        let contract_code = get_code(state_update, account_id, Some(account.code_hash))?;
+        if let Some(code) = contract_code {
+            // account storage usage should be larger than code size
+            let code_len = code.code.len() as u64;
+            debug_assert!(account_storage_usage > code_len);
+            account_storage_usage -= code_len;
```
Comment (evgenykuzyakov) [inline runtime/runtime/src/actions.rs:461]: Considering previous issue with state size due to delete key action, we should use `saturated_sub` to be safe

File: core/primitives/src/version.rs [hunk 2]
```diff
@@ -94,21 +94,23 @@ pub enum ProtocolFeature {
...
-pub const PROTOCOL_VERSION: ProtocolVersion = 42;
+pub const PROTOCOL_VERSION: ProtocolVersion = 43;
 
 /// Current latest nightly version of the protocol.
 #[cfg(feature = "nightly_protocol")]
 pub const PROTOCOL_VERSION: ProtocolVersion = 106;
 
 lazy_static! {
-    static ref STABLE_PROTOCOL_FEATURES_TO_VERSION_MAPPING: HashMap<ProtocolFeature, ProtocolVersion> =
+    pub static ref STABLE_PROTOCOL_FEATURES_TO_VERSION_MAPPING: HashMap<ProtocolFeature, ProtocolVersion> =
```
Comment (evgenykuzyakov) [inline core/primitives/src/version.rs:109]: This is not needed. nightly merges stable features in

File: runtime/runtime/src/actions.rs [hunk 3]
```diff
@@ -445,7 +446,28 @@ pub(crate) fn action_delete_account(
     result: &mut ActionResult,
     account_id: &AccountId,
     delete_account: &DeleteAccountAction,
+    current_protocol_version: ProtocolVersion,
 ) -> Result<(), StorageError> {
+    if current_protocol_version
+        >= STABLE_PROTOCOL_FEATURES_TO_VERSION_MAPPING[&ProtocolFeature::DeleteActionRestriction]
```
Comment (evgenykuzyakov) [inline runtime/runtime/src/actions.rs:452]:
```suggestion
        >= PROTOCOL_FEATURES_TO_VERSION_MAPPING[&ProtocolFeature::DeleteActionRestriction]
```

File: core/primitives/src/version.rs (hunk 2 above)
Comment (bowenwang1996) [reply core/primitives/src/version.rs:109]: Good point.

File: core/primitives/src/errors.rs [hunk 4]
```diff
@@ -677,7 +679,8 @@ impl Display for ActionErrorKind {
                 write!(f, "An new action receipt created during a FunctionCall is not valid: {}", e)
             }
             ActionErrorKind::InsufficientStake { account_id, stake, minimum_stake } => write!(f, "Account {} tries to stake {} but minimum required stake is {}", account_id, stake, minimum_stake),
-            ActionErrorKind::OnlyImplicitAccountCreationAllowed { account_id } => write!(f, "CreateAccount action is called on hex-characters account of length 64 {}", account_id)
+            ActionErrorKind::OnlyImplicitAccountCreationAllowed { account_id } => write!(f, "CreateAccount action is called on hex-characters account of length 64 {}", account_id),
+            ActionErrorKind::DeleteAccountWithLargeState { account_id } => write!(f, "The state of account {} is too large and therefore cannot be deleted", account_id)
```
Comment (matklad) [inline core/primitives/src/errors.rs:683]:
```suggestion
            ActionErrorKind::DeleteAccountWithLargeState { account_id } => write!(f, "The state of account {} is too large and therefore cannot be deleted", account_id),
```

File: neard/res/genesis_config.json [hunk 5]
```diff
@@ -1,239 +1,239 @@
 {
-  "protocol_version": 42,
-  "genesis_time": "1970-01-01T00:00:00.000000000Z",
-  "chain_id": "sample",
-  "genesis_height": 0,
-  "num_block_producer_seats": 50,
+  "protocol_version": 43, 
+  "genesis_time": "1970-01-01T00:00:00.000000000Z", 
+  "chain_id": "sample",
```
Comment (matklad) [inline neard/res/genesis_config.json:4]: trailing ws?

File: runtime/runtime/src/actions.rs [hunk 6]
```diff
@@ -445,7 +445,28 @@ pub(crate) fn action_delete_account(
     result: &mut ActionResult,
     account_id: &AccountId,
     delete_account: &DeleteAccountAction,
+    current_protocol_version: ProtocolVersion,
 ) -> Result<(), StorageError> {
+    if current_protocol_version
+        >= PROTOCOL_FEATURES_TO_VERSION_MAPPING[&ProtocolFeature::DeleteActionRestriction]
```
Comment (matklad) [inline runtime/runtime/src/actions.rs:451]: unrelated to PR, but feels like we can have a nicer API: `current_protocol_version.has_feature(ProtocolFeature::DeleteActionRestriction)`

File: neard/res/genesis_config.json (hunk 5 above)
Comment (bowenwang1996) [reply neard/res/genesis_config.json:4]: hmm this file is auto generated

File: runtime/runtime/src/actions.rs (hunk 6 above)
Comment (bowenwang1996) [reply runtime/runtime/src/actions.rs:451]: That is a good idea!

File: runtime/runtime/src/actions.rs [hunk 7]
```diff
@@ -458,7 +457,7 @@ pub(crate) fn action_delete_account(
             // account storage usage should be larger than code size
             let code_len = code.code.len() as u64;
             debug_assert!(account_storage_usage > code_len);
```
Comment (evgenykuzyakov) [inline runtime/runtime/src/actions.rs:459]: We don't run `debug_assert` in prod, right? So it's only for tests

**Review by evgenykuzyakov — APPROVED** [review]

File: runtime/runtime/src/actions.rs (hunk 7 above)
Comment (bowenwang1996) [reply runtime/runtime/src/actions.rs:459]: yes

**Pull request merged**